lib/code-caller-docker.js

```diff
@@ -0,0 +1,783 @@
+          this.outputBoth = data.errorData.outputBoth;
+        }
+      }
+    } catch (e) {
+      err = new Error('Error decoding DockerCaller JSON: ' + e.message);
+    }
+    this.state = WAITING;
+    if (err) {
+      this._callCallback(err);
+    } else {
+      if (data.functionMissing) {
+        this._callCallback(new FunctionMissingError('Function not found in module'));
+      } else {
+        this._callCallback(null, data.data, data.output);
+      }
+    }
+    this.debug('exit _callIsFinished()');
+  }
+
+  /**
+   * Will prepare this worker to be completely disposed of. This will kill
+   * the Docker container, clear the timeout, unmount the mounted directory
+   * if needed, and finally remove the mounted directory.
+   *
+   * This function SHOULD NOT THROW as we're not guaranteed that anyone will be
+   * able to catch the error. This makes a best-effort attempt to clean up all
+   * the resources used by this caller, but it'll simply ignore any errors it
+   * encounters.
+   */
+  async _cleanup() {
+    this.debug('enter _cleanup()');
+    if (this.timeoutID) {
+      this._clearTimeout();
+    }
+    if (this.container) {
+      try {
+        await this.container.kill();
+      } catch (e) {
+        logger.error(e);
+      }
+    }
+    // Note that we can't safely do any of this until the container is actually dead
+    if (this.hostDirectory) {
+      try {
+        await this.removeBindMount(this.hostDirectory.path);
+      } catch (e) {
+        // Probably not mounted; swallow error
+      }
+      try {
+        await this.hostDirectory.cleanup();
+      } catch (e) {
+        logger.error(e);
+      }
+    }
+    this.debug('exit _cleanup()');
+  }
+
+  _errorData() {
+    const errForStack = new Error();
+    return {
+      state: this.state,
+      containerIsNull: this.container == null,
+      callbackIsNull: this.callback == null,
+      timeoutIDIsNull: this.timeoutID == null,
+      outputBoth: this.outputBoth,
+      stack: errForStack.stack,
+      lastCallData: this.lastCallData,
+    };
+  }
+
+  /**
+   * @param {string} msg The message to log
+   */
+  _logError(msg) {
+    this.debug('enter _logError()');
+    const errData = this._errorData();
+    logger.error(msg, errData);
+    this.debug('exit _logError()');
+    return false;
+  }
+
+  /**
+   * Checks if the caller is ready for a call to call().
+   *
+   * @returns {boolean}
+   */
+  _checkReadyForCall() {
+    if (!this.container) {
+      return this._logError(
+        `Not ready for call, container is not created (state: ${String(this.state)})`
+      );
+    }
+    if (!this.coursePath) {
+      return this._logError(
+        `Not ready for call, course was not set (state: ${String(this.state)})`
+      );
+    }
+    return true;
+  }
+
+  /**
+   * Checks that the caller is in a good state.
+   *
+   * @param {CallerState[]} [allowedStates]
+   */
+  _checkState(allowedStates) {
```

Comment:
Should this also check the mutex states?

Reply:
I'm not sure which state I'd check for?

Reply:
I was thinking that the mutex should only be held in certain states? For example, is it possible for us to be in `state = IN_CALL` while still holding the `ensureChildMutex` mutex? If not, then we could assert that only certain combinations of state+mutex are possible. Of course, if we can have any combination of `state` and mutex then there is nothing to check.

Reply:
We decide mutexes are a good idea here.